src/object/cli_obj.c

```diff
@@ -1295,8 +1295,12 @@ dc_obj_fetch(tse_task_t *task)
 		D_GOTO(out_task, rc = -DER_INVAL);
 
 	rc = dc_tx_check(args->th, false, &epoch);
-	if (rc)
-		goto out_task;
+	if (rc) {
+		if (rc != -DER_INVAL)
+			goto out_task;
+		epoch = daos_ts2epoch();
+		D_DEBUG(DB_IO, "set epoch "DF_U64"\n", epoch);
```

Comment:
I was thinking that for fetch or obj_list/enumerate, the request needs not to be dispatched at server-side, moreover for EC fetch the request will be sent to multiple targets at client-side, so should set the epoch (same value for requests to different tgts) at client-side.
But I may mis-understood it?

Reply:
In theory, if the client and the server have never communicated before, then we cannot get ordering conclusion by comparing the client generated HLC/epoch with server side existing HLC/epoch. So generating HLC/epoch by server for fetch/list/query cases is more reasonable.

The most confused case may be for fetching data from EC object.

I am thinking that we should send the fetch RPC to one shard (in spite of leader or not), then such shard will forward the fetch RPC to other related shards. Different from update case, the first shard does not need to wait the forwarding RPC's reply, instead, the bulk transfer will between the client and related shards directly. So the overhead of fetching EC data from multiple shards will be half of RPC round trip.

On the other hand, if we make the client to dispatch the fetching EC RPC to multiple shards, even though the client itself can generate the epoch/HLC, it is still possible that the different shards may return different versions of EX data. I think that is wrong.

Reply:
> I am thinking that we should send the fetch RPC to one shard (in spite of leader or not), then such shard will forward the fetch RPC to other related shards.

If we need to ensure different clients get same result for fetch, seems should make fetch RPC goes to leader shard and forward to other shards? If fetch RPC first goes to different shard, then may with different epoch/HLC so may get different result.
And in case of leader switch, if the new leader did not communicate with old leader before, seems still possible cause different clients get different result? (for example first fetch served by old leader, and 2nd fetch served by new leader).

> On the other hand, if we make the client to dispatch the fetching EC RPC to multiple shards, even though the client itself can generate the epoch/HLC, it is still possible that the different shards may return different versions of EX data. I think that is wrong.

just wondering that as long as the fetching EC RPCs all with same epoch/HLC (set by client before dispatch), how could it happen that "the different shards may return different versions of EX data"?

Reply:
Assume that the epoch_A for fetch RPC is generated by the client and dispatched to both server_B and server_C (for reading EC object). The server_B receives the fetch RPC firstly, at that time, the server_C is handling new update request (from other client) which does not arrive at the server_B yet, so the server_B will reply the client without server_C's latest update. After that, the server_C has finished the update which becomes committable, and then the client's fetch RPC arrives at the server_C, then the server_C will reply the client with the latest update. That means the data fetched from the server_B and the server_C belong to the different update, right?

Reply:
right ... But that kind of race condition seems also cannot fully avoid even if the client fetch RPC firstly goes to leader shard and then dispatch.
To resolve it, will need to compare each shards' fetch's reply's epoch/HLC (if partial mismatch then retry). That seems needed for either client-sponsored or server-sponsored.

> I am thinking that we should send the fetch RPC to one shard (in spite of leader or not), then such shard will forward the fetch RPC to other related shards. Different from update case, the first shard does not need to wait the forwarding RPC's reply, instead, the bulk transfer will between the client and related shards directly. So the overhead of fetching EC data from multiple shards will be half of RPC round trip.

the server still need to wait dispatched req's completion before replying client RPC request, looks same with update case.

Client-sponsored dispatch for EC fetch looks better for performance (for normal case), sorry I am still not clear why must firstly goes to server and dispatch...

Reply:
If the fetching of EC object crosses multiple shards, should we guarantee that the data from multiple shards belong to the same update? Means whether allow the client to see partial update or not. For transaction semantics, it should NOT, but it is not required by POSIX.

We need to consider that well. My original idea is based on the assumption of guarantee that.

Reply:
just to be clear, by discussion with Fanyong and Johann, for fetch/obj_list we can assign the epoch/HLC at server-side for now, as currently only with replica obj.

Later for EC obj, we can change it in future that dispatch to different tgts shards (and assign the epoch/HLC) at client-side. 

thanks